Use the real timeToWork values in the top graph on the statistics page

voidus opened the pull request:
This includes a lot of refactoring.
https://github.com/Mayflower/PHProjekt/pull/341 should be merged before this.

Commit:
Moved the graph to a separate class

Commit:
Moved some intermediate values to separate functions

Commit:
add timecard model for reusability purposes

Commit:
use timecard model in statistics month graph

Commit:
add baseclass to the month graph to not depend on the html markup

Commit:
fix scoping errors in month graph

Commit:
put month graph into contentpane to make it swapable

Commit:
Moved remaining intermediate values to separate functions

They might be needed in multiple places.

Commit:
Actually call the function that renders the upper left rectangle

Commit:
We use single quotates for strings

Commit:
added small comment

Commit:
remove commented-out code

Commit:
slight clarification of comment

Commit:
Monthgraph always shows current month.

We used the current month in some places anyways, so let's remove this until we
can really support it.

Commit:
Added getWorkBalanceByDay to the timecard client model

Commit:
Introduce method object that renders the day blocks

See http://www.refactoring.com/catalog/replaceMethodWithMethodObject.html for an
explanation of a method object

Commit:
Try workBalanceByDay before using getMonthList

This will give us the minutes the user should work if she has a contract
configured. If not, it will give us an error, so we'll have to fall back to the
monthList.

Commit:
Fix typo bug that broke workBalanceByDayAction

Commit:
Merge remote-tracking branch 'exi/timehelperFixes' into realTimeToWork

* exi/timehelperFixes:
  fix timehelper dateToJsDate function because js months start with month 0...

Commit:
Added a geometry helper

This is a place for geometric calculations that is reusable between the
dayblockrenderer and the toworkrenderer

Commit:
Move some logic to separate functions

Commit:
Moved the overtime label calc to a separate function

Commit:
Extract method to render the vertical blue todayMarker

Commit:
Added TimeToWorkRenderer that renders the green line

This allows us to delete the old code. It still uses hard-coded 450minutes/day.

Commit:
Use the real minutesToWork values

exi commented on phprojekt/htdocs/phpr/Statistics/WorktimeMonthGraph.js:
```diff
@@ -0,0 +1,266 @@
+        }
+    });
+
+    var TimeToWorkRenderer = declare(null, {
```
I am a bit confuse as to why this is a declared object although it only has one function that is called like a function and seems to work like a function. does the declaration has any benefit except making the call (new Renderer().render()) more complicated?
I would put all the content into a regular function with a normal function scopy and inline help functions so it is clear how it is used.
Or am i missing something?
Same thing applies to the minutesBookedBlockRenderer

voidus replied:
Fixed

Commit:
Check for this._destroyed after all deferreds

Commit:
use single quotes instead of double quotes

Commit:
put import arguments on separate lines

Commit:
use correct quotation marks

Commit:
Encapsulate the temporary variables in a function instead of a method object.

With javascripts anonoymous functions and closures, we don't need method
objects. Instead, we can just declare the function and define it in an anonymous
function, that encapsulates the temporary variables.

This has the advantage that the code that uses the method is not aware of the
way it is built.

Commit:
use a function instead of a method object for the timeToWorkRenderer

Commit:
Move helper creation inside _renderTodayMarker

voidus commented on phprojekt/htdocs/phpr/Statistics/WorktimeMonthGraph.js:
```diff
@@ -0,0 +1,273 @@
+        };
+
+        var verticalX2 = function(d, i) {
+            return (i) * (barPadding + helper.barWidth());
```
(i)(i)

voidus replied:
fixed

voidus commented on phprojekt/htdocs/phpr/Statistics/WorktimeMonthGraph.js:
```diff
@@ -0,0 +1,273 @@
+        var dayEntries,
+            helper;
+
+        var renderHorizontalLines = function(svgData) {
```
function renderHorizontalLines(svgData) would be nicer

voidus replied:
fixed

voidus commented on phprojekt/htdocs/phpr/StatisticsView.js:
```diff
@@ -1,140 +1,13 @@
-], function(lang, declare, domAttr, locale, all, Widget, Templated, WidgetsInTemplate, api, timehelper,
-            templateString) {
+    'phpr/Statistics/WorktimeMonthGraph',
+], function(declare, Widget, Templated, WidgetsInTemplate, templateString) {
```
one item per line

voidus replied:
fixed

Commit:
Remove useless parentheses

Commit:
Use function foo() instead of var foo = function()

Commit:
print a single item per line in import lists

Commit:
Merge remote-tracking branch 'mayflower/next' into realTimeToWork

* mayflower/next:
  Remove the contact association from projects
  Always read the database file in the project migration
  Remove debug println and unnecessary line break
  remove dead code
  remove jshint reserved words
  make day blocks nicer
